functions/auth.js

```diff
@@ -19,3 +30,164 @@ exports.passwordsMatch = (submittedPassword, actualPassword) =>
+
+  // TODO: configure this in cloud config and/or firestore or similar
+  const scope = "";
+
+  const redirectUri = authClient.authorizationCode.authorizeURL({
+    redirect_uri: authCodeReturnUri,
+    scope,
+  });
+
+  functions.logger.log("Redirecting to:", redirectUri);
+  res.redirect(redirectUri);
+});
+
+/**
+ * Exchanges a given auth code passed in the 'code' URL query parameter for a Firebase auth token.
+ * The request also needs to specify a 'state' query parameter which will be checked against the 'state' cookie.
+ */
+exports.connectI4AOAuthHandler = functions.https.onRequest(async (req, res) => {
+  const { venueId } = req.query;
+
+  if (!AUTH_ORIGIN) {
+    throw new HttpsError("internal", "auth.origin is not configured properly");
+  }
+
+  if (!checkIfValidVenueId(venueId)) {
+    throw new HttpsError("invalid-argument", "venueId is not a valid venue id");
+  }
+
+  const authConfig = await fetchAuthConfig(venueId);
+
+  const {
+    i4aApiKey,
+    i4aOAuthUserInfoUrl,
+    i4aGetUserMeetingInfoUrl,
+  } = authConfig;
+
+  const authClient = createOAuth2Client(authConfig);
+
+  const { code: authCode } = req.query;
+
+  functions.logger.log("Received auth code:", authCode);
+  const results = await authClient.authorizationCode.getToken({
+    code: authCode,
+    redirect_uri: "http://localhost:3000/in/foo", // TODO: what do we want to use here?
+  });
+  functions.logger.log("Auth code exchange result received:", results);
+
+  const { access_token: accessToken } = results;
+
+  // Retrieve the user's I4A User ID
+  const { id: i4aUserId } = await getJson(i4aOAuthUserInfoUrl, {
+    Authorization: `Bearer ${accessToken}`,
+  });
+
+  functions.logger.log("I4A User ID:", i4aUserId);
+
+  if (!i4aUserId) {
+    // TODO: redirect to some kind of '500 error' page?
+    res.redirect("/in/TODO/error");
+    return;
+  }
+
+  // TODO: configure this in cloud config and/or firestore or similar
+  const meetingIdsToCheck = [131, 136];
+  const eventIdsToCheck = [504, 506, 554];
+
+  const checkedMeetingResult = await postJson(i4aGetUserMeetingInfoUrl, {
+    apiKey: i4aApiKey,
+    ams_id: i4aUserId,
+  });
+
+  functions.logger.log("Checked Meeting Result:", checkedMeetingResult);
+
+  const {
+    email: emailRaw,
+    meeting_ids: meetingIds = [],
+    event_ids: eventIds = [],
+  } = checkedMeetingResult;
+
+  const registeredMeetings = meetingIds.filter((meetingId) =>
+    meetingIdsToCheck.includes(meetingId)
+  );
+
+  const registeredEvents = eventIds.filter((eventId) =>
+    eventIdsToCheck.includes(eventId)
+  );
+
+  const isRegistered =
+    registeredMeetings.length > 0 && registeredEvents.length > 0;
+
+  functions.logger.log("Registered:", {
+    isRegistered,
+    registeredMeetings,
+    registeredEvents,
+  });
+
+  if (!isRegistered || !emailRaw) {
+    // TODO: redirect to some kind of 'not allowed' page
+    res.redirect("/in/TODO/not-registered");
+    return;
+  }
```

Comment:
[TODO] redirect back to the platform and show some kind of 'not allowed' page, that also prompts them to go and register/buy a ticket. We could potentially also send them straight to the buy a ticket page, but that might be confusing for them?